platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java

```diff
@@ -203,14 +204,39 @@ public Set<Object> getFrom() {
     public void setFromAnyValue(Object object) {
         // Currently we support just FileCollection for Groovy assign, so first try to cast to FileCollection
         FileCollectionInternal fileCollection = Cast.castNullable(FileCollectionInternal.class, Cast.castNullable(FileCollection.class, object));
+        checkNoSelfReferences(fileCollection);
+
+        setFrom(Cast.castNullable(FileCollection.class, object));
+    }
+
+    /**
+     * Verifies there is no self-referencing on value assignment.
+     * <p>
+     * Cover cases like: a += b, a = (a + b), a -= b, a = (a - b)
+     */
+    private void checkNoSelfReferences(FileCollectionInternal value) {
+        value.visitStructure(new FileCollectionStructureVisitor() {
+            private boolean isSubtracting = false;
 
-        // Don't allow a += b or a = (a + b), this is not support
-        fileCollection.visitStructure(new FileCollectionStructureVisitor() {
             @Override
-            public boolean startVisit(FileCollectionInternal.Source source, FileCollectionInternal fileCollection) {
+            public boolean startVisit(Source source, FileCollectionInternal fileCollection) {
                 if (DefaultConfigurableFileCollection.this == fileCollection) {
-                    throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Use the from() method to add to a ConfigurableFileCollection.");
+                    if (isSubtracting) {
+                        throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Avoid usage of the minus() or minus-assign operators.");
+                    } else {
+                        throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Use the from() method to add to a ConfigurableFileCollection.");
+                    }
+                }
+
+                if (fileCollection instanceof SubtractingFileCollection) {
+                    isSubtracting = true;
```

Comment:
❓ Should we set `isSubtracting = false` when we're done with traversing?

Reply:
My thought was to have a one-way switch, only to be able to give a better error message once we are inside a subtracting operation.

If we want to track when we are out, we'd need a counter at the very least. Otherwise, a second subtraction would trip the logic on the way out.